docs/cxx-proposal/cxx.rst

```diff
@@ -0,0 +1,523 @@
+really think of having one object, this choice makes it impossible to allow defining new schemas in Python.
+The same consequence applies to allowing for new schemas to be defined in Swift.)
+Ensuring a system that does not leak memory, and that also keeps both objects alive
+as long as either side (C++ or the bridged language) is, simply put, challenging.
+
+With all that as a preamble, here is our proposed solution for C++.
+
+- A new instance of a schema object is created by a call to ``new``.
+- All schema objects have protected destructors.  Given a raw pointer to
+  a schema object, client code may not directly invoke the ``delete`` operator,
+  but may write ::
+
+    Clip* c = new Clip();
+    ...
+    c->possibly_delete();    // returns true if c was deleted
+
+- The OTIO C++ API uses raw pointers exclusively in all its
+  function signatures (e.g. property access functions, property modifier functions, constructors, etc.).
+- Schema objects prevent premature destruction of schema instances they are interested
+  in by storing them in variables of type ``SerializableObject::Retainer<T>`` where ``T``
+  is of type ``SerializableObject`` (or derived from it).
+
+For example: ::
+
+  class ExtendedEffect : public Effect {
+  public:
+     ...
+     MediaReference* best() const {
+         return _best;
+     }
+
+     void set_best(MediaReference* best) {
+         _best = best;
+     }
+
+     MediaReference* best_or_other() {
+         return _best ? _best : some_other_reference();
+     }
+
+ private:
+   Retainer<MediaReference> _best;
+ };
+
+In this example, the ``ExtendedEffect`` schema has a property named ``best`` that must be
+a ``MediaReference``.  To indicate that it needs to retain its instance, the schema stores
+the property not as a raw pointer, but using the ``Retainer`` structure.
+
+Nothing special needs to be done for the reading/writing code, and there is automatic two-way
+conversion between ``Retainer<MediaReference>`` and ``MediaReference*`` which keeps the code
+simple.  Even testing if the property is set (as ``best_or_other()`` does) is done as
+if we were using raw pointers.
+
+The implementation of all this works as follow:
+
+- Creating a new schema instance starts the instance with an internal count of 0.
+- Putting a schema instance into a ``Retainer`` object increases the count by 1.
+- Destroying the retainer object or reassigning a new instance to it decreases the
+  count by 1 of the object if any in the retainer object.  If this causes the count
+  to reach zero, the schema instance is destroyed.
+- The ``possibly_delete()`` member function of ``SerializableObject*`` checks that
+  the count of the instance is zero, and if so deletes the object in question.
+- An ``any`` instance holding a ``SerializableObject*`` actually holds a
+  ``Retainer<SerializableObject>``.  That is, blind data safely retains schema instances.
+
```

Comment:
What's the mechanism for safety under concurrency? On the face of it, if I return a pointer, and Retainer the return value, isn't there a window where the pointer could be deleted before it is Retainer'd? Or is the idea that objects subject to this regime are owned at a higher scope - the schema from which the object was brokered holds the ultimate reference that in the usual case prevents this? Secondarily, I imagine the reference count is fenced and/or atomic?